nixos/modules/config/ipa.nix

```diff
@@ -0,0 +1,241 @@
+    SASL_NOCANON    on
+
+    URI ldaps://${cfg.server}
+    BASE ${cfg.basedn}
+    TLS_CACERT /etc/ipa/ca.crt
+  '';
+  nssDb = pkgs.runCommand "ipa-nssdb" { buildInputs = [ pkgs.nss.tools ]; } ''
+    set -x
+    mkdir -p $out
+    < /dev/urandom tr -dc '[:print:]' | head -c 40 > $out/pwdfile ||:
+    chmod 600 $out/pwdfile
+    certutil -d $out -N -f $out/pwdfile
+    chmod 644 $out/*.db
+    certutil -d $out -A -f $out/pwdfile -n "${cfg.realm} IPA CA" -t CT,C,C -i ${cfg.certificate}
+  '';
+
+in {
+  options = {
+    ipa = {
+      enable = mkEnableOption "FreeIPA domain integration";
+
+      certificate = mkOption {
+        type = types.package;
+        description = ''
+          IPA server CA certificate.
+
+          Use `nix-prefetch-url http://$server/ipa/config/ca.crt` to
+          obtain the file and the hash.
+        '';
+      };
+
+      domain = mkOption {
+        type = types.str;
+        example = "in.corp.com";
+        description = "Domain of the IPA server.";
+      };
+
+      realm = mkOption {
+        type = types.str;
+        example = "IN.CORP.COM";
+        description = "Kerberos realm.";
+      };
+
+      server = mkOption {
+        type = types.str;
+        example = "ipa.in.corp.com";
+        description = "IPA Server hostname.";
+      };
+
+      basedn = mkOption {
+        type = types.str;
+        example = "dc=in,dc=corp,dc=com";
+        description = "Base DN to use when performing LDAP operations.";
+      };
+
+      offlinePasswords = mkOption {
+        type = types.bool;
+        default = true;
+        description = "Whether to store offline passwords when the server is down.";
+      };
+
+      dyndns = {
+        enable = mkOption {
+          type = types.bool;
+          default = true;
+          description = "Whether to enable FreeIPA automatic hostname updates.";
+        };
+
+        interface = mkOption {
+          type = types.str;
+          example = "eth0";
+          default = "*";
+          description = "Network interface to perform hostname updates through.";
+        };
+      };
+
+      chromiumSupport = mkOption {
+        type = types.bool;
+        default = true;
+        description = "Whether to whitelist the FreeIPA domain in Chromium.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = sssd.enable;
+        message = "sssd must be enabled through `services.sssd.enable` for FreeIPA integration to work.";
+      }
+      {
+        assertion = ntp.enable;
+        message = "ntpd must be enabled through `services.ntp.enable` for FreeIPA integration to work.";
+      }
+      {
+        assertion = !krb5.enable;
+        message = "krb5 must be disabled through `krb5.enable` for FreeIPA integration to work.";
+      }
+      {
+        assertion = !ldap.enable;
+        message = "ldap must be disabled through `users.ldap.enable` for FreeIPA integration to work.";
+      }
+    ];
```

Comment:
Just turn them on automatically in config

Reply:
I left the two 'negative' assertions for `krb5` and `ldap`.